lib/Optimizer/IROptimizer.cpp

```diff
@@ -344,44 +344,30 @@ static void sinkTensorViews(IRFunction &M) {
 static void deleteDeadAllocs(IRFunction &M) {
   auto &instrs = M.getInstrs();
 
-  llvm::SmallVector<Instruction *, 16> erasedInstructions{};
-
-  // Remove all unused tensorviews.
-  for (auto &I : instrs) {
-    if (isa<TensorViewInst>(I) && I.getNumUsers() == 0) {
-      erasedInstructions.push_back(&I);
+  // Remove all unused tensor views tracking back their dependencies, which are
+  // in a topological order.
+  // Note that this should precede to remove dependencies on allocs.
+  for (auto it = instrs.rbegin(); it != instrs.rend();) {
+    // Remember the current instruction and advance the iterator.
+    auto *I = &*it++;
+    if (isa<TensorViewInst>(I) && I->getNumUsers() == 0) {
+      // Remove a tensor view. It may make other tensor views preceding it
+      // eligible for a removal as well.
+      M.eraseInstruction(I);
     }
   }
 
-  for (auto I : erasedInstructions) {
-    M.eraseInstruction(I);
-  }
-
-  erasedInstructions.clear();
+  // Remove all of unused allocs and their corresponding deallocs.
+  for (auto it = instrs.rbegin(); it != instrs.rend();) {
+    auto *I = &*it++;
 
-  // Remove all of the DeallocActivationInst that close unused allocs.
-  for (auto &I : instrs) {
-    const auto *DA = dyn_cast<const DeallocActivationInst>(&I);
+    const auto *DA = dyn_cast<const DeallocActivationInst>(I);
     if (DA && DA->getAlloc()->getNumUsers() < 2) {
-      erasedInstructions.push_back(&I);
-    }
-  }
-
-  for (auto I : erasedInstructions) {
-    M.eraseInstruction(I);
-  }
-  erasedInstructions.clear();
-
-  // Remove the unused allocs.
-  for (auto &I : instrs) {
-    if (isa<AllocActivationInst>(&I) && I.getNumUsers() < 2) {
-      erasedInstructions.push_back(&I);
+      M.eraseInstruction(I);
+    } else if (isa<AllocActivationInst>(I) && I->getNumUsers() < 2) {
```

Comment:
I think that this condition `I->getNumUsers() < 2` can be formulated more precisely in this re-written version of the code. Since we perform a reverse iteration and first remove deallocs, when we reach the alloc it would have no uses at all, or?

And again, please explain before the loop why we iterate in a reverse order and that it means that deallocs are eliminated before their respective allocs.

Reply:
The original code removes deallocs first without explanation, But I think that it is to prevent `DeallocActivationInst::getAlloc()` from returning erased allocs. 
What do you mean by formulating the condition?

Reply:
I mean that once the the dealloc is erased, the getNumUsers() of the alloc becomes 0. So, the condition may be written as:

```c++
     if (DA && DA->getAlloc()->getNumUsers() < 2) {
       M.eraseInstruction(I);
       continue;
     }
     if (isa<AllocActivationInst>(I) && !I->hasUsers()) {
       M.eraseInstruction(I); 
     }
```

Also note, that my example does not use `else`, as LLVM coding style prefers avoiding else if it possible by using early exist, etc.